cpp/include/rapidsmpf/streaming/core/memory_reserve_or_wait.hpp

```diff
@@ -0,0 +1,196 @@
+     *
+     * This coroutine submits a memory reservation request and suspends until
+     * either sufficient memory becomes available or no progress is made within the
+     * configured timeout.
+     *
+     * If the request cannot be satisfied within the timeout, the function attempts
+     * to reserve the requested memory by allowing overbooking. This guarantees
+     * forward progress at the cost of potentially exceeding the configured memory
+     * limits.
+     *
+     * @param size Number of bytes to reserve.
+     * @param future_release_potential Estimated number of bytes the requester may release
+     * in the future, used as a heuristic when selecting which eligible request to satisfy
+     * first.
+     * @return A pair consisting of:
+     *   - A `MemoryReservation` representing the allocated memory.
+     *   - The number of bytes by which the reservation overbooked the available memory.
+     *     This value is zero if no overbooking occurred.
+     *
+     * @throws std::runtime_error If shutdown occurs before the request can be processed.
+     */
+    coro::task<std::pair<MemoryReservation, std::size_t>> reserve_or_wait_or_overbook(
+        std::size_t size, std::size_t future_release_potential
+    );
+
+    /**
+     * @brief Returns the number of pending memory reservation requests.
+     *
+     * It may change concurrently as requests are added or fulfilled.
+     *
+     * @return The number of outstanding reservation requests.
+     */
+    [[nodiscard]] std::size_t size() const;
+
+    /**
+     * @brief Returns the number of iterations performed by `periodic_memory_check()`.
+     *
+     * This counter is incremented once per loop iteration inside
+     * `periodic_memory_check()`, and can be useful for diagnostics or testing.
+     *
+     * @return The total number of memory-check iterations executed so far.
+     */
+    [[nodiscard]] std::size_t periodic_memory_check_counter() const;
+
+  private:
+    /**
+     * @brief Represents a single memory reservation request.
+     *
+     * A `Request` is inserted into a sorted container and processed by
+     * `periodic_memory_check()`.
+     */
+    struct Request {
+        /// @brief The number of bytes requested.
+        std::size_t size;
+
+        /// @brief Estimated number of bytes expected to be released in the future.
+        std::size_t future_release_potential;
+
+        /// @brief Monotonically increasing identifier used to preserve submission order.
+        std::uint64_t sequence_number;
+
+        /// @brief Queue into which a reservation is pushed once the request is satisfied.
+        coro::queue<MemoryReservation>& queue;
+
+        /// @brief Ordering by `size` and `sequence_number` (ascending).
+        friend bool operator<(Request const& a, Request const& b) {
+            return std::tie(a.size, a.sequence_number)
+                   < std::tie(b.size, b.sequence_number);
```

Comment:
When looking at the `periodic_memory_check` impl, I feel like the sequence_number is immaterial. We take the smallest/ range of leq requests based on size, and then try to find the one max future_release_potential. So, I feel like sequence_number order is not really required, and comparing on just the size is sufficient.

Reply:
It is used as tie breaker so that each request is unique, it is also used on timeout:
```c++
        // Among the smallest requests, pick the one with the largest
        // future_release_potential. If multiple requests tie, we pick the oldest one,
        // since the set is ordered by size and then sequence_number (ascending).
        auto it = std::ranges::max_element(
            std::ranges::subrange(first, same_size_end),
            std::less<>{},
            &Request::future_release_potential
        );
```